contracts/epochs/FlowEpoch.cdc

```diff
@@ -109,46 +128,551 @@ pub contract FlowEpochs {
+
+        // Get nodeinfo for all the nodes
+        // get all the collector and consensus nodes
+        // to initialize the QC and DKG
+        for id in ids {
+            let nodeInfo = FlowIDTableStaking.NodeInfo(nodeID: id)
+
+            nodeInfoArray.append(nodeInfo)
+
+            if nodeInfo.role == NodeRole.Collector.rawValue {
+                collectorNodeIDs.append(nodeInfo.id)
+            }
+
+            if nodeInfo.role == NodeRole.Consensus.rawValue {
+                consensusNodeIDs.append(nodeInfo.id)
+            }
+        }
+        
+        // Organize the collector nodes into clusters
+        let collectorClusters = self.createCollectorClusters(nodeIDs: collectorNodeIDs)
+
+        // Start QC Voting with the supplied clusters
+        self.QCAdmin.startVoting(clusters: collectorClusters)
+
+        // Start DKG with the consensus nodes
+        self.DKGAdmin.startDKG(nodeIDs: consensusNodeIDs)
+
+        // Initialze the metadata for the next epoch
+        // QC and DKG metadata will be filled in later
+        let proposedEpochMetadata = EpochMetadata(counter: self.proposedEpochCounter(),
+                                                seed: randomSource,
+                                                startView: self.epochMetadata[self.currentEpochCounter]!.endView + UInt64(1),
+                                                endView: self.epochMetadata[self.currentEpochCounter]!.endView + self.configurableMetadata.numViewsInEpoch,
+                                                stakingEndView: self.epochMetadata[self.currentEpochCounter]!.endView + self.configurableMetadata.numViewsInStakingAuction,
+                                                collectorClusters: collectorClusters,
+                                                clusterQCs: [],
+                                                dkgKeys: [])
+
+        self.epochMetadata[self.proposedEpochCounter()] = proposedEpochMetadata
+
+        self.currentEpochPhase = EpochPhase.EPOCHSETUP
+
+        emit EpochSetup(counter: proposedEpochMetadata.counter,
+                        nodeInfo: nodeInfoArray, 
+                        firstView: proposedEpochMetadata.startView,
+                        finalView: proposedEpochMetadata.endView,
+                        collectorClusters: collectorClusters,
+                        randomSource: randomSource,
+                        DKGPhase1FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + self.configurableMetadata.numViewsInDKGPhase,
+                        DKGPhase2FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + (2 as UInt64 * self.configurableMetadata.numViewsInDKGPhase),
+                        DKGPhase3FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + (3 as UInt64 * self.configurableMetadata.numViewsInDKGPhase))
```

Comment:
I believe these are off by 1
```suggestion
                        DKGPhase1FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + self.configurableMetadata.numViewsInDKGPhase -1 as UInt64,
                        DKGPhase2FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + (2 as UInt64 * self.configurableMetadata.numViewsInDKGPhase) - 1 as UInt64,
                        DKGPhase3FinalView: proposedEpochMetadata.startView + self.configurableMetadata.numViewsInStakingAuction + (3 as UInt64 * self.configurableMetadata.numViewsInDKGPhase) -1 as UInt64)
```

Reply:
yes, I believe you are correct. Good catch!